### PR TITLE
feat: add multisig example

### PR DESCRIPTION
**Pull request by kratico**
Add a multi-signature transfer example.

The example does the following
- derive a multi-signature address for Alice, Bob and Charlie with 2 as a threshold
- Alice transfers some balance to the multi-signature address
- Alice creates a proposal to transfer to Dave from the multi-signature address
- Bob approves the proposal to transfer to Dave from the multi-signature address
- log Dave's new account balance to the console

### CODE REVIEW

File: examples/multisig_transfer.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        value: sender.sign(message),
+      };
+    },
+    createWatchHandler,
+  });
+
+const _getTimepoint = () => {
+  return C.readEntry(config, "Multisig", "Multisigs", [
+    ABC_MULTISIG_ADDRESS,
+  ]);
```
Comment (kratico) [inline examples/multisig_transfer.ts:93]: The timepoint is required for any approval after the proposal creation

image

File: frame_metadata/Key.ts [hunk 2]
```diff
@@ -28,23 +28,39 @@ export function $storageKey(props: StorageKeyProps): $.Codec<unknown[]> {
...
-    ),
+  const $keys = [...Array(keyCodecs.length + 1).keys()].reduce(
+    (acc, i) => {
+      const codec = $.tuple(
+        ...keyCodecs.slice(0, i).map(($key, i) =>
+          H[(props.storageEntry as M.MapStorageEntryType).hashers[i]!].$hash($key)
+        ),
+      );
+      const keySize = keyCodecs.slice(0, i).reduce(
+        (acc, $key) => acc + ($key._metadata![1] as number),
```
Comment (kratico) [inline frame_metadata/Key.ts:39]: This is the key param size
There may be a better API to get this info

File: frame_metadata/Key.ts [hunk 3]
```diff
@@ -28,23 +28,39 @@ export function $storageKey(props: StorageKeyProps): $.Codec<unknown[]> {
...
-    ),
+  const $keys = [...Array(keyCodecs.length + 1).keys()].reduce(
+    (keys, i) => {
+      const codec = $.tuple(
+        ...keyCodecs.slice(0, i).map(($key, i) =>
+          H[(props.storageEntry as M.MapStorageEntryType).hashers[i]!].$hash($key)
+        ),
+      );
+      const keySize = keyCodecs.slice(0, i).reduce(
+        (size, $key) => size + ($key._metadata![1] as number),
```
Comment (kratico) [inline frame_metadata/Key.ts:39]: This is the key param size
There may be a better API to get this info

File: examples/multisig_transfer.ts (hunk 1 above)
Comment (kratico) [reply examples/multisig_transfer.ts:93]: Depends on https://github.com/paritytech/capi/issues/294

File: examples/multisig_transfer.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+if (!hostname || !portRaw) {
+  throw new Error("Must be running inside a test ctx");
+}
+
+import { createKeyMulti } from "../deps/polkadot/util-crypto.ts";
+import * as C from "../mod.ts";
+import * as T from "../test_util/mod.ts";
+import * as U from "../util/mod.ts";
+
+type KeyringPair = typeof T.alice;
```
Comment (harrysolovay) [inline examples/multisig_transfer.ts:13]: Please delete this line and instead specify the `Uint8Array` and `C.M.SignExtrinsic` types where applicable.

File: examples/multisig_transfer.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    }))
+    .watch(function(status) {
+      console.log(status);
+      if (C.TransactionStatus.isTerminal(status)) {
+        this.stop();
+      }
+    });
+}
+
+interface CreateOrApproveProposalProps {
```
Comment (harrysolovay) [inline examples/multisig_transfer.ts:86]: Given that this interface isn't reused, should we inline this constraints of `createOrApproveMultisigProposal`? Ie.

```diff
function createOrApproveMultisigProposal(
- { sender, maybe_timepoint }: CreateOrApproveProposalProps,
+ publicKey: Uint8Array,
+ sign: C.M.SignExtrinsic,
+ maybeTimepoint?: bigint,
) {
```

File: examples/multisig_transfer.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    })
+      .signed((message) => ({
+        type: "Sr25519",
+        value: sender.sign(message),
+      }));
+  };
+  const withoutActualWeight = createEx().extrinsic;
+  const weight = C.payment.queryInfo(T.polkadot)(withoutActualWeight)
+    .access("weight");
+  C.Z.call(weight, (d) => {
```
Comment (harrysolovay) [inline examples/multisig_transfer.ts:140]: Not sure if this serves the example (as much as debugging the example). Can we get rid?

File: rpc/client.ts [hunk 7]
```diff
@@ -53,10 +53,11 @@ export class Client<
     } else if (e.id) {
       const pendingCall = this.pendingCalls[e.id];
       if (!pendingCall) {
-        console.log({ e });
+        console.log("!pendingCall for", { e });
         // TODO: pipe error to listeners and message the likely cause,
         //       a duplicate client.
-        throw new Error();
+        // throw new Error();
+        return;
```
Comment (harrysolovay) [inline rpc/client.ts:60]: Is not throwing a good idea? No pending call would signal faulty behavior.

File: test_util/local.ts [hunk 8]
```diff
@@ -53,6 +53,9 @@ export class LocalClientEffect extends Z.Effect<LocalClient, PolkadotBinNotFound
   }
 
   private async createClient(): Promise<LocalClient> {
+    // TODO: remove logging
+    console.log("creating client...");
```
Comment (harrysolovay) [inline test_util/local.ts:57]: Added `nice console messaging from test ctx / spawned Polkadot processes` to our ongoing list in #215.

**Review by harrysolovay — CHANGES_REQUESTED** [review]

File: examples/multisig_transfer.ts (hunk 6 above)
Comment (kratico) [reply examples/multisig_transfer.ts:140]: right now, the `weight` is hardcoded, I'll comment it out until there is a better way to inject effects in the extrinsic params

File: rpc/client.ts (hunk 7 above)
Comment (kratico) [reply rpc/client.ts:60]: will revert

File: examples/multisig_transfer.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+// Create proposal
+U.throwIfError(
+  await createOrApproveMultisigProposal({
+    publicKey: T.alice.publicKey,
+    sign: C.Z.call(0, () =>
+      (message) => ({
+        type: "Sr25519",
+        value: T.alice.sign(message),
+      })),
```
Comment (kratico) [inline examples/multisig_transfer.ts:35]: @harrysolovay I found a weird issue.
If a plain function is used this doesn't work.
It seems that the result of the first `sign()` call is cached so on the 2nd extrinsic (when bob approves the proposal) the signature check fails

Comment (harrysolovay) [reply examples/multisig_transfer.ts]: For now, let's just use the test ctx and call each with `run` (wihout first binding to an env).

**Review by harrysolovay — APPROVED** [review]

**Pull request merged**